src/v/kafka/server/handlers/fetch.cc

```diff
@@ -671,6 +672,477 @@ class parallel_fetch_plan_executor final : public fetch_plan_executor::impl {
     }
 };
 
+/**
+ * A fetch exeutor that relies on notifications from raft to determine when to
+ * query partitions. Rather than querying all partitions at a set polling
+ * interval like the `parallel_fetch_plan_executor`.
+ */
+class nonpolling_fetch_plan_executor final : public fetch_plan_executor::impl {
+    // Passed from the coordinator shard to fetch workers.
+    // Contains either references to objects local to the fetch worker shard or
+    // copies of data from the coordinator shard.
+    struct shard_local_fetch_context {
+        bool foreign_read;
+        size_t bytes_left;
+        size_t min_bytes;
+        std::optional<model::timeout_clock::time_point> deadline;
+        std::vector<ntp_fetch_config> requests;
+        server& srv;
+        cluster::partition_manager& mgr;
+        ss::abort_source& as;
+    };
+
+    // Returned by fetch workers to the coordinator shard.
+    struct worker_result {
+        std::vector<read_result> read_results;
+        std::vector<size_t> errored_partitions;
+        size_t total_size;
+    };
+
+public:
+    nonpolling_fetch_plan_executor()
+      : _last_result_size(ss::smp::count, 0) {}
+
+    /**
+     * Executes the supplied `plan` until `octx.should_stop_fetch` returns true.
+     */
+    ss::future<> execute_plan(op_context& octx, fetch_plan plan) final {
+        co_await _as.start();
```

Comment:
Why doesn't this use `octx.rctx.abort_source()` as parent?

Reply:
Great point, it would remove the need for the ` ssx: allow for abort_source to be created without a parent` too. Switching to it now.